drivers/flash/disk_flash.c

```diff
@@ -0,0 +1,344 @@
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (len_write > data->sector_buf_size - read_padding) {
+			len_write = data->sector_buf_size - read_padding;
+		}
+
+		memcpy(&dst_buf[write_cursor], &data->sector_buf[read_padding], len_write);
+		write_cursor += len_write;
+	}
+
+release_device:
+	release_device(dev);
+	return ret;
+}
+
+static int disk_flash_write(const struct device *dev, off_t offset, const void *src, size_t len)
+{
+	int ret = 0;
+	const uint8_t *src_buf = src;
+	const struct disk_flash_config *cfg = dev->config;
+	struct disk_flash_data *data = dev->data;
+	uint32_t read_cursor = 0;
+
+	acquire_device(dev);
+
+	ret = check_boundary(dev, offset, len);
+	if (ret != 0) {
+		goto release_device;
+	}
+
+	if (data->disk_sector_size == 0) {
+		ret = -EINVAL;
+		goto release_device;
+	}
+
+	while (read_cursor < len) {
+		uint32_t write_cursor = cfg->disk_offset + offset + read_cursor;
+		uint32_t write_sector = write_cursor / data->disk_sector_size;
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - read_cursor);
+
+		if (write_sector < 0 || write_sector >= data->disk_total_sector_cnt) {
+			LOG_ERR("Sector out of bounds");
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (disk_access_read(cfg->disk_name, data->sector_buf, write_sector, 1)) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (len_write > data->sector_buf_size - write_padding) {
+			len_write = data->sector_buf_size - write_padding;
+		}
+
+		memcpy(&data->sector_buf[write_padding], &src_buf[read_cursor], len_write);
+
+		if (disk_access_write(cfg->disk_name, data->sector_buf, write_sector, 1)) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+		read_cursor += len_write;
+	}
+
+release_device:
+	release_device(dev);
+	return ret;
+}
+
+static int disk_flash_erase(const struct device *dev, off_t offset, size_t size)
+{
+	int ret = 0;
+	const struct disk_flash_config *cfg = dev->config;
+	struct disk_flash_data *data = dev->data;
+	uint32_t read_cursor = 0;
+
+	acquire_device(dev);
+
+	ret = check_boundary(dev, offset, size);
+	if (ret != 0) {
+		goto release_device;
+	}
+
+	if (data->disk_sector_size == 0) {
+		ret = -EINVAL;
+		goto release_device;
+	}
+
+	while (read_cursor < size) {
+		uint32_t write_cursor = cfg->disk_offset + offset + read_cursor;
+		uint32_t write_sector = write_cursor / data->disk_sector_size;
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (size - read_cursor);
+
+		if (write_sector < 0 || write_sector >= data->disk_total_sector_cnt) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (disk_access_read(cfg->disk_name, data->sector_buf, write_sector, 1)) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (len_write > data->sector_buf_size - write_padding) {
+			len_write = data->sector_buf_size - write_padding;
+		}
+
+		memset(&data->sector_buf[write_padding],
+			cfg->flash_parameters.erase_value,
+			len_write);
```

Comment:
You could do that once, before entering the loop.

Reply:
Since the erase supports arbitrary size and does not need to be aligned with the sector of the disk partial sectors need to be erased. Therefore I don't currently see a possibility to do this outside the loop.